concepts/src/controllers/pagination.ts

```diff
@@ -0,0 +1,102 @@
+import { URL, URLSearchParams } from "url";
+import { HttpError } from "./error";
+
+const limits = {
+  minSize: 1,
+  maxSize: 100,
+};
+
+const defaultPageSize = 10;
+
+export type PaginationQueryParameters = {
+  page?: number;
+  pageSize?: number;
+};
+
+export type PaginationResponse = {
+  pageSize: number;
+  totalPages: number;
+  totalResults: number;
+  prevPage?: string;
+  nextPage?: string;
+};
+
+export const paginationElasticBody = ({
+  page,
+  pageSize,
+}: PaginationQueryParameters): { from: number; size: number } => {
+  const size = Math.min(pageSize ?? defaultPageSize, limits.maxSize);
+  return {
+    size,
+    from: ((page ?? 1) - 1) * size,
+  };
+};
+
+const parsePaginationQueryParameters = (
+  params: URLSearchParams
+): PaginationQueryParameters => {
+  const page = parseNumberParam("page", params);
+  const pageSize = parseNumberParam("pageSize", params);
+
+  if (page !== undefined && page < 1) {
+    throw new HttpError({
+      status: 400,
+      label: "Bad request",
```

Comment:
nit: these are usually capitalised

```suggestion
      label: "Bad Request",
```